### PR TITLE
[GEOT-7502] Optimize FlatGeobuf DataStore for raw Query with startIndex

### PR DESCRIPTION
**Pull request by bjornharrtell**
[GEOT-7502](https://osgeo-org.atlassian.net/browse/GEOT-7502) [Powered by Pull Request Badge](https://pullrequestbadge.com/?utm_medium=github&utm_source=geotools&utm_campaign=badge_info)

  

# Checklist

- [x] I have read the [contribution guidelines](https://github.com/geotools/geotools/blob/main/CONTRIBUTING.md).
- [x] I have sent a [Contribution Licence Agreement](https://docs.geotools.org/latest/developer/procedures/contribution_license.html) (not required for small changes, e.g., fixing typos in documentation).
- [x] First PR targets the `main` branch (backports managed later; ignore for branch specific issues).
- [x] Avoid [Java 9+ split packages](http://tutorials.jenkov.com/java/modules.html#split-packages-not-allowed).
- [ ] All the build checks are green ([see automated QA checks](https://docs.geotools.org/latest/developer/conventions/code/qa.html)).

For core and extension modules:

- [x] New unit tests have been added covering the changes.
- [x] [Documentation](https://github.com/geotools/geotools/tree/main/docs) has been updated (if change is visible to end users).
- [x] There is an issue in [GeoTools Jira](https://osgeo-org.atlassian.net/projects/GEOT) (except for changes not visible to end users). 
- [ ] Commit message(s) must be in the form ``[GEOT-XYZW] Title of the Jira ticket``.
- [ ] Bug fixes and small new features are presented as a single commit.
- [x] The commit targets a single objective (if multiple focuses cannot be avoided, each one is in its own commit, and has a separate ticket describing it).

### CODE REVIEW

**Comment by bjornharrtell** [comment]
@aaime when you got time I would appreciate if you can sanity check this.

File: modules/unsupported/flatgeobuf/src/test/java/org/geotools/data/flatgeobuf/FlatGeobufDataStoreTest.java [hunk 1]
```diff
@@ -903,6 +903,24 @@ public void readCountriesFids() throws IOException {
...
+        query.setMaxFeatures(2);
+        query.setStartIndex(10);
+        SimpleFeatureCollection featureCollection = featureSource.getFeatures(query);
+        try (SimpleFeatureIterator it = featureCollection.features()) {
+            SimpleFeature f1 = it.next();
+            assertEquals("countries.107", f1.getID());
+            SimpleFeature f2 = it.next();
+            assertEquals("countries.108", f2.getID());
+            boolean hasNext = it.hasNext();
+            assertFalse(hasNext);
```
Comment (aaime) [inline modules/unsupported/flatgeobuf/src/test/java/org/geotools/data/flatgeobuf/FlatGeobufDataStoreTest.java:920]: Against a database one would need to use a order by to get a predictable output sequence, but it's very easy to imagine a file read to have a predictable order instead. Just trying to make sure.

**Review by aaime — CHANGES_REQUESTED** [review]
What's missing here is the implementation of "canOffset -> true" in the FlatGeoBufFeatureSource.

File: modules/unsupported/flatgeobuf/src/test/java/org/geotools/data/flatgeobuf/FlatGeobufDataStoreTest.java (hunk 1 above)
Comment (bjornharrtell) [reply modules/unsupported/flatgeobuf/src/test/java/org/geotools/data/flatgeobuf/FlatGeobufDataStoreTest.java:920]: Hmm yes, and FlatGeobuf doesn't allow for any mutation so ordering can be assumed to be static.

**Comment by bjornharrtell** [comment]
> What's missing here is the implementation of "canOffset -> true" in the FlatGeoBufFeatureSource.

Hmm ok. I just have be sure something doesn't think it can do more than it can - that is, a FlatGeobuf can only be read with an offset if it has an index and there is no other (spatial) filter.

**Comment by bjornharrtell** [comment]
@aaime I think I see what you mean... I've tried to get my head around it but as of yet I'm failing to understand how to implement. It seems a bit deep of a rabbit hole (to me) because it works togheter with sorting (which does make sense in the general case) but I'm not sure how to special case / override that.

**Comment by bjornharrtell** [comment]
The current implementation isn't effective. It is not possible to implement in a straight forward was due to inability to optimize this case for specific queries, see https://github.com/geotools/geotools/pull/4593.

Added a failing test what I think is what should be the correct result. The fail seems to have something to do with the default sorting that is applied in the non native case and should be looked into.

**Comment by bjornharrtell** [comment]
Merged with https://github.com/geotools/geotools/pull/4593 and https://github.com/geotools/geotools/pull/4594 and reenabled the optimization and looks like it working.

**Comment by aaime** [comment]
With https://github.com/geotools/geotools/pull/4593 merged, would you update this one? Or close?

**Comment by jodygarnett** [comment]
@bjornharrtell I opened up this ticket https://osgeo-org.atlassian.net/browse/GEOT-7509 to cover doc change missed by this PR.

**Pull request closed without merging**